crates/astria-config/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{
+    parse_macro_input,
+    Item,
+};
+
+#[proc_macro_attribute]
+pub fn astria_config(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let item_ast = parse_macro_input!(item as Item);
+
+    let Item::Struct(ref item_struct) = item_ast else {
+        panic!("the astria_config proc macro can only be called on structs");
+    };
+
+    let struct_name = item_struct.ident.clone();
+
+    // Generates an vec of token-tree like structures of the atomicity we care about
+    let struct_fields = item_struct
+        .fields
+        .iter()
+        .map(|x| {
+            let field_ident = x.ident.clone().unwrap();
+            let field_ty = x.ty.clone();
+            let field_vis = x.vis.clone();
+
+            (field_ident, field_ty, field_vis)
+        })
+        .collect::<Vec<_>>();
+
+    // Check if log is a valid member of the config, otherwise panic
+    let true = struct_fields
+        .iter()
+        .map(|x| x.0.to_string().to_lowercase())
+        .collect::<Vec<_>>()
+        .contains(&"log".to_string())
+    else {
+        panic!("config struct does not contain a log element, which is necessary");
+    };
+
+    // Check if the macro itself has a valid attribute
+    let attr_str = attr.to_string().to_lowercase();
+    match attr_str.as_str() {
+        "conductor" | "composer" | "sequencer" | "sequencer_relayer" => {}
+        _ => panic!(
+            "Invalid attribute {attr_str}: please select a valid astria service to generate \
+             config for"
+        ),
+    };
+
+    let prefix = format!("ASTRIA_{}_", attr_str.to_uppercase());
+    let test_prefix = format!("TESTTEST_{}", prefix);
+
+    // let struct_field_ast = struct_fields
+    //     .iter()
+    //     .map(|(id, ty, vis)| {
+    //         quote! {
+    //             #vis #id: #ty
+    //         }
+    //     })
+    //     .collect::<Vec<_>>();
+
+    let code_gen_ast = quote! {
+        const ENV_PREFIX: &str = #prefix;
+
+        use serde::{Serialize, Deserialize};
+
+        // Adding serde dependencies on top of the old struct
+        #[derive(Debug, Deserialize, Serialize)]
+        #[serde(deny_unknown_fields)]
+        #item_ast
+
+        pub fn get() -> Result<#struct_name, figment::Error> {
+            #struct_name::from_environment(ENV_PREFIX)
+        }
+
+        impl #struct_name {
+            fn from_environment(env_prefix: &str) -> Result<Config, figment::Error> {
+                figment::Figment::new()
+                    .merge(figment::providers::Env::prefixed("RUST_").split("_").only(&["log"]))
+                    .merge(figment::providers::Env::prefixed(env_prefix))
+                    .extract()
+            }
+        }
+
+        #[cfg(test)]
```

Comment:
Test should probably never be made part of generated code. Should one of these fail, it's difficult for users to find the problem.

Reply:
I agree with this statement but disagree about these specific tests because they are sanity checks. If these are failing, then they did not follow the spec. I don't think these specific tests being super visible adds to that. I was going to write documentation that basically pasted these so that people can refer to them from the cargo docs. But I think having these tests everywhere is unnecessary clutter.

Reply:
Update: Had a brief conversation with @SuperFluffy about this, and we came up with a tentative in between solution